### DISCUSSION TITLE
[FEATURE REQUEST] Create the output directory if it doesn't exist

### DISCUSSION THREAD

**Discussion by jarmos-san** [tectonic-typesetting/tectonic#891] [Ideas]
## What's the Issue

Running the following command:

```console
tectonic foo.tex --output bar
```
results in the error mentioned below because the directory `bar` doesn't exists yet.

```console
error: output directory "bar" does not exist
```

## Possible Solution

A simple fix to the above workflow would be to simply create the input directory `bar` if it doesn't exist already. In doing so, it would relieve the user of any manual tasks no matter how trivial.

It's a subtle feature but I'm pretty sure will be well-received by many users out there (unless I'm missing out on something obvious).

**Comment by pkgw** [comment]
I'm going to convert this to a Discussion, since I'm curious to hear feedback from folks about what they think the most desirable behavior is here.

**Comment by pkgw** [comment]
As per the above: what do people think of the above? I tend to think that it's better for tools to be cautious and error out in circumstances like these, but I don't have a super-strong opinion here.

**Reply by jarmos-san** [reply]
Agreed about making tools cautious rather than perform some actions under the hood without the user's notice. Unfortunately, I'm not a Rust dev & as such I can't contribute solid ideas to the feature's implementation detail. But I suggest designing the feature on the lines of how the following command works:

```console
mkdir -p foo
```
... which basically translates to: "_make the parent directory `foo` if they do not already exist_".

That was my two cents but I look forward to hearing more expert opinion on it (again I would like to emphasise on the fact that I'm no Rust dev).

**Reply by pkgw** [reply]
Yeah, the technical implementation would be straightforward if we wanted to offer this kind of behavior — just a few lines of code, I expect. So it's more about the policy decision of what kind of behavior is preferable.

**Reply by ratmice** [reply]
The only issue I can think of is in the shared directories like `/tmp`, with the *the directory must already exist* behavior, when you go to create the directory for the output, if the directory already exists e.g. created by another user, the user would notice when they themselves create the directory, however if the user is expecting tectonic to create the directory but it already exists created by another user.

In that case, I suppose it's possible there is some kind of issue when tectonic does multiple passes over the output/rereading it as input when shell-escape is enabled?
Edit: I'm not sure you could get access to shell-escape but at the very least unintended input through additional passes...

I suppose its a somewhat far-fetched corner case.

**Comment by arcstur** [comment]
Maybe it coulde be an option, like putting `--create-dir` would create it if does not exists, but I really think that the default behavior should be to give an error.

**Comment by Starrah** [comment]
I just want to show my personally agree on the feature being added, since it is very convenient for the users. I think making directory is a somehow of no harm in most circumstances, since users can always delete the directory along with the generated file after finding their mistake. Creating directory won't cause any permenant loss, so I think its advantage outperforms drawback for creating directory automatically. Just my personal opinion.

**Comment by cdelledonne** [comment]
I also second this feature request.

I don't see any major risks in not being cautious here, and nobody has posted any significant one in this discussion (perhaps I'm missing something?) As @Starrah, I also believe the advantage of this feature would be far more compelling than any possible drawback.

On top of that, the disadvantage in not having this feature implemented explodes in situations where source files are organized in a more complex folder structure. The output directory tree of one of my projects looks like the following, and creating all the subdirectories manually when the output directory does not exists is somewhat painful, and also not very maintainable in automated build environments/CI pipelines:
```
build
├── chapters
│   ├── appendix
│   │   ├── apps.aux
│   │   └── arch.aux
│   ├── arch
│   │   └── index.aux
│   ├── background
│   │   └── index.aux
│   ├── conclusion
│   │   └── index.aux
│   ├── doa
│   │   └── index.aux
│   ├── intro
│   │   └── index.aux
│   ├── meta
│   │   ├── acks.aux
│   │   ├── cv.aux
│   │   ├── publications.aux
│   │   ├── summary.aux
│   │   └── title.aux
│   └── netstack
│       └── index.aux
├── dissertation.aux
├── dissertation.bbl
├── dissertation.bcf
├── dissertation.blg
├── dissertation.log
├── dissertation.pdf
├── dissertation.run.xml
├── dissertation.synctex.gz
└── dissertation.toc
```
